_posts/2023-11-28-shrinking-the-scheduler.markdown

```diff
@@ -0,0 +1,81 @@
+
+The scheduler has to be in the TCB for availability but ideally it shouldn't be for confidentiality or integrity.
+For event groups, this was mostly true.
+A malicious or compromised scheduler could set or clear event notifications, causing things to wake spuriously or not, but you could at least mitigate this by also communicating any key state separately from the event group and using it just to tell you to check something else.
+
+
+Message queues complicated this picture.
+The scheduler could see all messages in message queues.
+Worse, message queues were reachable from the thread structures and so a compromise of the scheduler in one thread could leak or corrupt any data in any in-flight message queues.
+This put the scheduler in the TCB for confidentiality and integrity of the contents of message queues.
+
+Futexes can be used to implement a rich set of locks and can also be used to implement message queues and event groups.
+This is precisely what [PR139](https://github.com/microsoft/cheriot-rtos/pull/139) is doing.
+
+Message queues are interesting because they have a variety of threat models.
+In some cases, they're used for passing messages between threads in the same compartment.
+In others, they're expected to be a trusted intermediary between two compartments in different trust domains.
+
+In the first case, both ends trust each other to follow the rules and so we can expose a message-queue library and invoke it without a cross-compartment call.
+This makes the overhead of sending and receiving in a message queue lower.
+You need a cross-compartment call only if the queue is transitioning from empty to non-empty, full to non-full, or if there is contention on sending or receiving, the cases where a thread in one endpoint may be sleeping.
+All other cases are a direct call to a shared library.
+
+In the second case, we need another compartment to ensure that the message-queue rules are followed, for example not overwriting arbitrary messages and ensuring that the producer and consumer counters are incremented correctly.
+The scheduler was doing this, but in the new version there is an option compartment that can fill this rôle.
+This compartment is completely unprivileged and compromising it will not impact the scheduler.
+It is also almost entirely stateless.
+Two different message queues are never visible to it at the same time, all communication between threads happens via message queue objects that are passed in (as sealed capabilities) from callers.
+The only state that it owns is the cached thread ID pointer returned from `thread_id_get_pointer()` by the scheduler.
+This means that, if you find a bug in the message queue code (entirely possible!) and force it to crash, it will not affect any other message queues, only yours.
+In particular, it won't leak data from message queues or corrupt their state.
```

Comment:
This paragraph seems to skip over what the compartment actually does.  Perhaps just add something simple like "This new compartment encapsulates message queue operations, with message queues exposed via sealed capabilities." after "rôle" (and removing "all communication ... callers", perhaps).